src/Symfony/Component/HttpKernel/Kernel.php

```diff
@@ -539,10 +539,16 @@ protected function initializeContainer()
             touch($oldContainerDir.'.legacy');
         }
 
-        $preload = $this instanceof WarmableInterface ? (array) $this->warmUp($this->container->getParameter('kernel.cache_dir'), $buildDir) : [];
+        $preload = $this instanceof WarmableInterface ? (array) $this->warmUp($buildDir, $buildDir) : [];
 
         if ($this->container->has('cache_warmer')) {
-            $preload = array_merge($preload, (array) $this->container->get('cache_warmer')->warmUp($this->container->getParameter('kernel.cache_dir'), $buildDir));
+            $cacheWarmer = $this->container->get('cache_warmer');
+
+            if ($this->container->getParameter('kernel.cache_dir') !== $buildDir) {
+                $cacheWarmer->enableOptionalWarmers();
+            }
+
+            $preload = array_merge($preload, (array) $cacheWarmer->warmUp($buildDir, $buildDir));
```

Comment:
because passing the cacheDir is highly confusing (which one should I use?), and warmers don't need to know where the cache-dir is when there is a build-dir

Reply:
I've reverted this part: passing the cacheDir actually forces warmers to opt-in for the build dir if they really want to support it
this is safer: not updated cache warmer will continue writing to the cache dir